adapters/silverpush/silverpush.go

```diff
@@ -0,0 +1,344 @@
+			siteCopy.Publisher = &openrtb2.Publisher{ID: impExt.PublisherId}
+		} else {
+			publisher := *siteCopy.Publisher
+			publisher.ID = impExt.PublisherId
+			siteCopy.Publisher = &publisher
+		}
+		req.Site = &siteCopy
+
+	} else if req.App != nil {
+		appCopy := *req.App
+		if appCopy.Publisher == nil {
+			appCopy.Publisher = &openrtb2.Publisher{ID: impExt.PublisherId}
+		} else {
+			publisher := *appCopy.Publisher
+			publisher.ID = impExt.PublisherId
+			appCopy.Publisher = &publisher
+		}
+		appCopy.Publisher = &openrtb2.Publisher{ID: impExt.PublisherId}
+		req.App = &appCopy
+
+	}
+
+	return nil
+}
+
+func impressionByMediaType(imp *openrtb2.Imp) ([]openrtb2.Imp, error) {
+
+	if imp.Banner == nil && imp.Video == nil {
+		return nil, &errortypes.BadInput{Message: "Invalid MediaType. SilverPush only supports Banner, Video"}
+	}
+	imps := make([]openrtb2.Imp, 0, 2)
+
+	if imp.Banner != nil {
+		impCopy := *imp
+		impCopy.Video = nil
+		imps = append(imps, impCopy)
+	}
+	if imp.Video != nil {
+		impCopy := *imp
+		impCopy.Banner = nil
+		imps = append(imps, impCopy)
+	}
+	return imps, nil
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}}
+	}
+
+	var bidResp openrtb2.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(internalRequest.Imp))
+
+	// overrride default currency
+	if bidResp.Cur != "" {
+		bidResponse.Currency = bidResp.Cur
+	}
+
+	for _, sb := range bidResp.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &sb.Bid[i],
+				BidType: getMediaTypeForImp(sb.Bid[i].ImpID, internalRequest.Imp),
+			})
+		}
+	}
+	return bidResponse, nil
+}
+
+// getMediaTypeForImp figures out which media type this bid is for.
+// SilverPush doesn't support multi-type impressions.
+// If both banner and video exist, take banner as we do not want in-banner video.
+func getMediaTypeForImp(impId string, imps []openrtb2.Imp) (mediaType openrtb_ext.BidType) {
+	mediaType = openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Banner != nil {
+				mediaType = openrtb_ext.BidTypeBanner
+			} else if imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+		}
```

Comment:
Warning - This is an anti-pattern. This code assumes that for multi-format bid request, mediatype will be set to type banner by default. I would recommend to set the mediaType based on bidder response. Your bidder should be able to set openrtb2.Bid.MType field and use that field here to get mediaTypeForImp. Even if your bidder does not support multi-format request, I would highly encourage you to make this change.

Reply:
Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)